Sistr cmd update 2018

jrober84 opened the pull request:
Significant changes to SISTR antigen biomarker and cgMLST database and processing of flagella antigens

apetkau commented on sistr/tests/test_cgmlst.py:
```diff
@@ -0,0 +1,39 @@
+'''
```
Also, I'm wondering how you were running the tests (or if you were making use of them)? I tried running `pytest` but I get a number of errors like:

```
ImportError while importing test module '/home/CSCScience.ca/apetkau/workspace/sistr_cmd/sistr/tests/test_hclust_cutree.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
sistr/tests/test_hclust_cutree.py:1: in <module>
    from sistr.src.cgmlst.extras.hclust_cutree import profiles_to_np_array, dist_matrix_hamming, complete_linkage, cutree, nr_profiles, expand_clusters_dataframe
sistr/src/cgmlst/extras/hclust_cutree.py:1: in <module>
    from fastcluster import linkage
E   ModuleNotFoundError: No module named 'fastcluster'

sistr/tests/test_writer_util.py:5
  /home/CSCScience.ca/apetkau/workspace/sistr_cmd/sistr/tests/test_writer_util.py:5: PytestCollectionWarning: cannot collect test class 'Test1' because it has a __init__ constructor
    class Test1:
```

jrober84 replied:
I have removed the duplicated test directory. At present I am not using the tests, they were designed by Peter. I think that we can add for a future issue adding full coverage of the SISTR code base by tests.

apetkau replied:
Thanks. That sounds good to me. Could you make that issue on github (to add tests for SISTR code base)?

jrober84 commented on sistr/data/antigens/fliC.fasta:
```diff
@@ -1,3 +1,1304 @@
-version https://git-lfs.github.com/spec/v1
-oid sha256:9e9e11cde490dadd450231398df83a20add73431e24ae397732e32d0801d7be3
-size 1076904
+>350|e,n,x,z15
```
Deleted now

jrober84 commented on sistr/sistr_cmd.py:
```diff
@@ -146,29 +148,83 @@ def merge_cgmlst_prediction(serovar_prediction, cgmlst_prediction):
+        shutil.rmtree(resource_filename('sistr', 'data/'))
+        os.mkdir(resource_filename('sistr', 'data/'))
+    if (not os.path.isfile(tmp_file)):
+        logging.error('Downloading databases failed, please check your internet connection and retry')
```
Added the download url to the file and to the download message

jrober84 commented on sistr/sistr_cmd.py:
```diff
@@ -146,29 +148,83 @@ def merge_cgmlst_prediction(serovar_prediction, cgmlst_prediction):
+    else:
+        logging.info('Downloading databases successful')
+        f = open(resource_filename('sistr', 'dbstatus.txt'),'w')
+        f.write("DB downloaded on : {}".format(datetime.today().strftime('%Y-%m-%d')))
```
Added the download url  to the file

jrober84 commented on sistr/sistr_cmd.py:
```diff
@@ -146,29 +148,83 @@ def merge_cgmlst_prediction(serovar_prediction, cgmlst_prediction):
+            f_out.close()
+            os.remove(fname)
+
+def setup_sistr_dbs(logging):
```
All of the files are in the download package so I have removed the stub directory

jrober84 commented on sistr/misc/reduce_to_centroid_alleles.py:
```diff
@@ -65,12 +65,17 @@ def find_all_centroid_alleles(marker_alleles, threads=1, word_size=28, cluster_t
     if threads == 1:
         logging.info('Running centroid allele finding in serial single-threaded mode')
         for marker_name, alleles in marker_alleles.items():
+            #print (marker_name)
```
Removed

jrober84 commented on sistr/misc/add_ref_genomes.py:
```diff
@@ -273,6 +274,7 @@ def write_serovar_and_spp_tables(outdir, df_serovar, predictions, genome_names):
     pred_genome_serovar = {}
     pred_genome_spp = {}
     for genome, prediction in zip(genome_names, predictions):
+        print(genome)
```
Removed the statements, they were for debugging

jrober84 commented on sistr/sistr_cmd.py:
```diff
@@ -290,10 +347,13 @@ def write_novel_alleles(cgmlst_results, output_path):
     init_console_logger(args.verbose)
     logging.info('Running sistr_cmd {}'.format(__version__))
+    if not os.path.isfile(resource_filename('sistr', 'dbstatus.txt')):
+        setup_sistr_dbs(logging)
```
Removed the passing of logging

jrober84 commented:
Thanks @apetkau  for the review, these were very helpful suggestions. I have fixed the handling of 0 fasta files in sistr_cmd in commit 3a81ab2 and resolved the warnings in 55077d9.

peterk87 commented:
Hi @apetkau and @jrober84 

Just to comment on the cgMLST ST and allele number issue @apetkau highlighted:

> It looks like the majority of the SISTR cgMLST sequence types will change (not sure if this is expected)?

When working on the cgMLST DB for v1.0.2, I found that curation of what alleles you allow into the cgMLST scheme to be paramount to ensuring the reproducibility of the cgMLST results over time especially when adding alleles from large numbers of genomes. The cgMLST Sequence Type (ST) is just a md5sum hash of a complete cgMLST allelic profile where the cgMLST allele numbers are just a md5sum hash of the underlying sequence of the allele. So the cgMLST ST is especially sensitive to slight changes to the allele sequences. 

For v1.0.2, I ended up looking through the MSAs for the [medoid](https://en.wikipedia.org/wiki/Medoid) alleles (previously called them centroid alleles erroneously) and removing anything that *visually* looked like it was introducing too much variation (i.e. throwing the alignment out of alignment). I re-ran the reduced set of alleles with `sistr_cmd` against a set of reference genomes to produce a new cgMLST DB. I wouldn't recommend doing something like this again since it's a very manual process. Maybe only using alleles that are okay by `sistr_cmd` v1.0.2 in the new DB with some quick checks to ensure that no new alleles are out of alignment with the existing alleles?

It may be a good idea to pass the cgMLST FASTA DB through something like [`db-check`](https://github.com/andersgs/db-check) to remove potentially problematic alleles.

Mostly unchanging cgMLST allele numbers and STs from v1.0.2 to v1.1.0 would be a great feature of `sistr_cmd`. In theory, the allele sequences found by SISTR between versions should be the same for each marker so the allele numbers should also be the same. But there's probably errors in v1.0.2 that have been fixed in v1.1.0 or slightly longer alleles that have been introduced for some markers that match better by `sistr_cmd`'s metrics. 

Another thing I was wondering about is whether allelic profile results between MacOS and Linux are comparable. That is, does the following function return the same value on different OSes and platforms (see [`allele_name` source code](https://github.com/phac-nml/sistr_cmd/blob/master/sistr/src/cgmlst/__init__.py#L20))?

```python
zlib.crc32(seq) & 0xffffffff
```

Hope that helps!

apetkau commented:
Thanks for the additional info @peterk87. Changing allele numbers due to corrected errors or longer longer alleles makes sense as to why the cgMLST sequence types would change. Especially since the number is so sensitive to differences (if any allele changes at all, the sequence type changes).

I agree it would be very nice to have the cgMLST allele numbers mostly consistent. But, I also realize that this could involve a lot of work. And our primary purpose for SISTR is still serotyping (rather than cgMLST). So, I would be okay if these change (so long as serotypes are consistent). When this is integrated into IRIDA, the changes will all appear under a brand new column in our metadata table anyways (for SISTR `1.1.0`), so results won't be mixed up.

However, many other people use SISTR as well. And I don't know how much they rely on the cgMLST numbers being consistent?

At a minimum, I think we should make it very clear that cgMLST types will change across versions of SISTR (e.g., in a `CHANGELOG.md` file, and in our release notes).

What are your thoughts @jrober84 ?

> Another thing I was wondering about is whether allelic profile results between MacOS and Linux are comparable. That is, does the following function return the same value on different OSes and platforms (see allele_name source code)?

I think CRC32 is used a lot for detecting errors in data (e.g., during transmission across networks). It's also used for error checking in **zip** archives. So, I suspect it should be consistent across different platforms (otherwise zipped files on Linux would break on MacOS or Windows). The Python docs for that function states:

>To generate the same numeric value across all Python versions and platforms, use crc32(data) & 0xffffffff. -  https://docs.python.org/3/library/zlib.html#zlib.crc32

This is exactly what you do, so I hope it should be consistent. Though, I wonder if there ever could be issues with different ways of encoding text (would this lead to different CRC32 values)?

jrober84 commented on sistr/src/cgmlst/extras/centroid_cgmlst_alleles.py:
```diff
@@ -5,8 +5,8 @@
-NT_TO_INT = {'A':1,'C':2,'G':3,'T':4}
-INT_TO_NT = {1:'A',2:'C',3:'G',4:'T'}
+NT_TO_INT = {'A':1,'C':2,'G':3,'T':4,'N':5}
+INT_TO_NT = {1:'A',2:'C',3:'G',4:'T',5:'5'}
```
Good catch @apetkau, I fixed this to be N instead of 5

apetkau replied:
Awesome. Thanks.

jrober84 commented on sistr/src/serovar_prediction/__init__.py:
```diff
@@ -508,9 +557,9 @@ def overall_serovar_call(serovar_prediction, antigen_predictor):
-            else:
-                if float(cgmlst_distance) <= CGMLST_DISTANCE_THRESHOLD:
-                    serovar_prediction.serovar = cgmlst_serovar
+            #else:
```
Removed

jrober84 commented on sistr/src/serovar_prediction/constants.py:
```diff
@@ -8,6 +8,8 @@
 CGMLST_SUBSPECIATION_DISTANCE_THRESHOLD = 0.9
 MASH_SUBSPECIATION_DISTANCE_THRESHOLD = 0.01
 
+SISTR_DB_URL = 'https://irida.corefacility.ca/downloads/sistr/database/SISTR_V_1.1_db.tar.gz'
```
I think if it is going to happen in a year or two then it is fine to leave it as it is now. I used figshare for mob-suite and they changed the url structure within six months. So there is no guarantee that figshare won't modify their urls earlier than us.

apetkau replied:
That makes sense. I did not realize FigShare have changed their URL structure in the past.

The switch is still a little ways away yet (there's still our IRIDA Galaxy toolshed which depends on that URL that we're slowly shifting tools out of as they get updated).

jrober84 commented:
Thanks @peterk87 and @apetkau, the change in sequence types is expected as @peterk87 said even a single allele change will change the ST type.  The focus from SISTR is on the serotype calls and from the perspective of the majority of users, they are not using the cgMLST information.  I agree with Aaron that we should make it clear that the cgMLST types can change between DB release versions and for now that is the best that we can do.

I have fixed the mash issue as it was the result of old data in the sketch which was no longer included in the database. I have added error checking to allow for mismatches between the mash sketch and the genomes-to-serovar table

apetkau commented:
Thanks for all the fixes @jrober84.

That sounds good about the cgMLST information changing. Sounds like we're all on the same page.

I'm going to spend a bit more time testing this out before merging, but it overall looks great to me :+1:

apetkau commented on sistr/src/serovar_prediction/__init__.py:
```diff
@@ -141,7 +143,12 @@ def search_for_wzx(self):
                 return
 
-            if top_result_length < 400:
+            if top_result_length < 300:
```
I realize it's a small thing, but I noticed that if `top_result_length == 300`, then it won't hit any of these `if` statements (and so won't be recorded as as missing)? Is this proper, it should you use `if top_result_length <= 300`?

apetkau commented on sistr/src/serovar_prediction/__init__.py:
```diff
@@ -161,11 +168,17 @@ def search_for_wzy(self):
                 return
 
-            if top_result_length < 600:
+            if top_result_length < 300:
```
Same comment here about `top_result_length`.

jrober84 replied:
Thanks for catching that, it has been fixed in commit 3ef9915..3925340

apetkau commented:
Everything looks great to me. Thanks so much @jrober84
